include/cuco/priority_queue.cuh

```diff
@@ -0,0 +1,239 @@
+#pragma once
+
+#include <cuco/allocator.hpp>
+#include <utility>
+#include <vector>
+
+#include <thrust/functional.h>
+
+namespace cuco {
+
+/*
+ * @brief A GPU-accelerated priority queue of key-value pairs
+ *
+ * Allows for multiple concurrent insertions as well as multiple concurrent
+ * deletions
+ *
+ * Current limitations:
+ * - Does not support insertion and deletion at the same time
+ *   - The implementation of the priority queue is based on
+ *     https://arxiv.org/pdf/1906.06504.pdf, which provides a way to allow
+ *     concurrent insertion and deletion, so this could be added later if useful
+ * - Capacity is fixed and the queue does not automatically resize
+ * - Deletion from the queue is much slower than insertion into the queue
+ *   due to congestion at the underlying heap's root node
+ *
+ * The queue supports two operations:
+ *   `push`: Add elements into the queue
+ *   `pop`: Remove the element(s) with the lowest (when Max == false) or highest
+ *        (when Max == true) keys
+ *
+ * The priority queue supports bulk host-side operations and more fine-grained
+ * device-side operations.
+ *
+ * The host-side bulk operations `push` and `pop` allow an arbitrary number of
+ * elements to be pushed to or popped from the queue.
+ *
+ * The device-side operations allow a cooperative group to push or pop from
+ * device code. These device side
+ * operations are invoked with a trivially-copyable device view,
+ * `device_mutable_view` which can be obtained with the host function
+ * `get_mutable_device_view` and passed to the device.
+ *
+ * @tparam T Type of the elements stored in the queue
+ * @tparam Compare Comparison operator used to order the elements in the queue
+ * @tparam FavorInsertionPerformance When true, insertion performance is
+ *         increased at the expense of deletion performance.
+ * @tparam Allocator Allocator defining how memory is allocated internally
+ */
+template <typename T,
+          typename Compare               = thrust::less<T>,
+          bool FavorInsertionPerformance = false,
+          typename Allocator             = cuco::cuda_allocator<char>>
+class priority_queue {
+  using int_allocator_type = typename std::allocator_traits<Allocator>::rebind_alloc<int>;
+
+  using t_allocator_type = typename std::allocator_traits<Allocator>::rebind_alloc<T>;
+
+  using size_t_allocator_type = typename std::allocator_traits<Allocator>::rebind_alloc<size_t>;
+
+  const int NodeSize = FavorInsertionPerformance ? 64 : 1024;
+
+ public:
+  /**
+   * @brief Construct a priority queue
+   *
+   * @param initial_capacity The number of elements the priority queue can hold
+   * @param alloc Allocator used for allocating device storage
+   */
+  priority_queue(size_t initial_capacity, Allocator const& alloc = Allocator{});
```

Comment:
Maybe we should pass a stream parameter to set `d_*` variables asynchronously.

Reply:
[4269e9c](https://github.com/NVIDIA/cuCollections/pull/105/commits/4269e9c4dce144ba95ed95dd98cb2713d77bc1ea)